src/smc-webapp/admin/site-settings.tsx

```diff
@@ -182,65 +198,202 @@ export class SiteSettings extends Component<{}, SiteSettingsState> {
+      (typeof valid == "function" && !valid(value))
+    ) {
+      return { backgroundColor: "red", color: "white" };
+    }
+    return {};
+  }
+
+  private on_change_entry(name, val?) {
+    const e = copy(this.state.edited);
+    e[name] = val ?? ReactDOM.findDOMNode(this.refs[name]).value;
+    return this.setState({ edited: e });
+  }
+
+  private render_row_entry_inner(name, value, valid, password): Rendered {
+    if (Array.isArray(valid)) {
+      return (
+        <Select
+          defaultValue={value}
+          onChange={val => this.on_change_entry(name, val)}
+          style={{ width: "100%" }}
+        >
+          {valid.map(e => (
+            <Select.Option value={e} key={e}>
+              {e}
+            </Select.Option>
+          ))}
+        </Select>
+      );
+    } else {
+      if (password) {
+        return (
+          <Input.Password
+            style={this.row_entry_style(value, valid)}
+            value={value}
+            visibilityToggle={true}
+            onChange={val => this.on_change_entry(name, val)}
+          />
+        );
+      } else {
+        return (
+          <Input
+            ref={name}
+            style={this.row_entry_style(value, valid)}
+            value={value}
+            onChange={() => this.on_change_entry(name)}
+          />
+        );
+      }
+    }
+  }
+
+  private render_row_entry(
+    name: string,
+    value: string,
+    password: boolean,
+    parsed_val?: string,
+    valid?: ConfigValid,
+    hint?: Rendered
+  ) {
     switch (name) {
       case "default_quotas":
       case "max_upgrades":
         return this.render_json_entry(name, value);
       default:
         return (
           <FormGroup>
-            <FormControl
-              ref={name}
-              type="text"
-              value={value}
-              onChange={() => {
-                const e = copy(this.state.edited);
-                e[name] = ReactDOM.findDOMNode(this.refs[name]).value;
-                return this.setState({ edited: e });
-              }}
-            />
-            {name === "version_recommended_browser"
-              ? this.render_version_hint(value)
-              : undefined}
+            {this.render_row_entry_inner(name, value, valid, password)}
+            <div style={{ fontSize: "90%", display: "inlineBlock" }}>
+              {this.render_row_version_hint(name, value)}
+              {hint}
+              {this.render_row_entry_parsed(parsed_val)}
+              {this.render_row_entry_valid(valid)}
+            </div>
           </FormGroup>
         );
     }
   }
 
-  render_row(name: string, value: string): Rendered | undefined {
-    if (value == null) {
-      value = site_settings_conf[name].default;
+  private render_default_row(name): Rendered | undefined {
+    const conf: Config = site_settings_conf[name];
+    return this.render_row(name, conf);
+  }
+
+  private render_extras_row(name): Rendered | undefined {
```

Comment:
@haraldschilly you don't need to do `Rendered| undefined` since Rendered **is** undefined | JSX.Element.  I.e., it allows for undefined already...